### PR TITLE
Push notifs Config

### PR DESCRIPTION
**Pull request by wthomasmiii**
# Overview

## TL;DR 
Addition of push notification to webhook endpoint.

### What does this change?

1. Adds helpers to perform push functions
    1a. Register Device
    1b. Unregister Device
    1c. Send Push Notification 

2. Adds sending the notifications for
    2a. [X] Send Connection Request
    2b. [X] Accept Connection Request
    2c. [X] Send VC
    2d. [ ] Accept VC
    2e. [X] Send VP
    2f.  [ ] Accept VP

 ### Testing
The following are testable functions of this update.

1. Registering a device with the tRPC registration call, where a profileId is got from context, and you pass in a deviceToken as an argument
2. Performing requests such as connection, VC, or VP sending/requests that trigger a push notification

### CODE REVIEW

File: services/learn-card-network/brain-service/src/helpers/connection.helpers.ts [hunk 1]
```diff
@@ -199,6 +208,14 @@ export const requestConnection = async (
 
     await source.relateTo({ alias: 'connectionRequested', where: { profileId: target.profileId } });
 
+    await SendPushNotification({
+        to: target.profileId,
+        message: `${source.displayName} has sent you a connection request!`,
+        title: 'New Connection Request',
+        url: `/connect/request?profileId=${source.displayName}&connect=true`,
```
Comment (gerardopar) [inline services/learn-card-network/brain-service/src/helpers/connection.helpers.ts:215]: replace `/connect/request?profileId=${source.displayName}&connect=true` with ``/contacts/requests?profileId=${source.displayName}&connect=true`` 🙂

**Review by gerardopar — CHANGES_REQUESTED** [review]
@wthomasmiii left some comments regarding the url redirects 😊

File: .prettierrc.json [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
-  "arrowParens": "avoid",
-  "overrides": [
-    {
-      "files": ".prettierrc",
-      "options": { "parser": "json" }
-    }
-  ]
+    "printWidth": 100,
+    "trailingComma": "es5",
+    "tabWidth": 2,
```
Comment (Custard7) [inline .prettierrc.json:4]: Why is this tabWidth changed?

File: services/learn-card-network/brain-service/src/helpers/connection.helpers.ts [hunk 3]
```diff
@@ -1,156 +1,162 @@
...
-        }),
-        source.relateTo({ alias: 'connectedWith', where: { profileId: target.profileId } }),
-        target.relateTo({ alias: 'connectedWith', where: { profileId: source.profileId } }),
-    ]);
+  await SendPushNotification({
+    to: source.profileId,
+    message: `${target.displayName} has accepted your connection request!`,
+    title: 'Connection Accepted',
+    actionType: 'redirect',
+    url: '/contacts',
```
Comment (Custard7) [inline services/learn-card-network/brain-service/src/helpers/connection.helpers.ts:123]: This is okay for now - ideally in the future we might want to refactor this to be a custom uri, like data: `lcn:redirect:JSON.stringify({ url: '/contacts' })` or something like that

File: services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+interface IPushRegistrationRequest {
+    profileId: string;
+    deviceToken: string;
+}
+
+export async function SendPushNotification(args: IPushNotificationRequest) {
+    try {
+        if (process.env.WEBHOOK_URL) {
```
Comment (Custard7) [inline services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts:16]: Have we set this in git secrets? If not, we should!

Should we rename it to NOTIFICATIONS_SERVICE_WEBHOOK_URL to disambiguate?

File: services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            const response = await fetch(process.env.WEBHOOK_URL, {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json',
+                    'apikey': `${process.env.APIKEY}`,
```
Comment (Custard7) [inline services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts:21]: Same here, and should we rename to NOTIFICATIONS_SERVICE_API_KEY to disambiguate?

File: services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            });
+            if (!response.ok) {
+                throw new Error(`HTTP error! status: ${response.status}`);
+            }
+            const data = await response.json();
+            return data;
+        }
+    } catch (error) {
+        console.error('Error fetching data:', error);
+        console.log(error);
```
Comment (Custard7) [inline services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts:41]: Should we remove the normal console log?

```suggestion
        console.error('Notifications Helpers - Send Push Notification:', error);
```

Comment (Custard7) [reply services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts:41]: Should we log extra information here for easy debuggin?

File: services/learn-card-network/brain-service/src/helpers/presentation.helpers.ts [hunk 7]
```diff
@@ -24,6 +25,14 @@ export const sendPresentation = async (
 
     await createSentPresentationRelationship(from, to, presentationInstance);
 
+    await SendPushNotification({
+        to: to.profileId,
+        message: `${from.displayName} has boosted you!`,
+        title: 'Boost Received',
+        url: `/notifications?claim=true`,
+        actionType: 'redirect',
```
Comment (Custard7) [inline services/learn-card-network/brain-service/src/helpers/presentation.helpers.ts:33]: I think @gerardopar also needs the uri embedded in this one too, yeah?

**Review by Custard7 — COMMENTED** [review]
I left a few comments - the main two that need to be addressed:

- The prettier config was changed, which also makes some of these commits to be a bit messy. Is there a reason for this? If not, let's revert the prettier settings, and re-prettify each of these files to eliminate the git-change noise.
- I believe we definitely need to pass uri in the boost notification, as that is the main credential sending function being used right now, but can verify with @gerardopar

File: services/learn-card-network/brain-service/src/helpers/presentation.helpers.ts (hunk 7 above)
Comment (wthomasmiii) [reply services/learn-card-network/brain-service/src/helpers/presentation.helpers.ts]: added

File: services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts (hunk 6 above)
Comment (wthomasmiii) [reply services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts]: removed standard console.log, replaced with more helpful error

File: services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts (hunk 5 above)
Comment (wthomasmiii) [reply services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts]: changed

File: services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts (hunk 4 above)
Comment (wthomasmiii) [reply services/learn-card-network/brain-service/src/helpers/notifications.helpers.ts]: changed

**Review by gerardopar — APPROVED** [review]

**Review by Custard7 — APPROVED** [review]
